packages/core/src/client.ts

```diff
@@ -151,89 +151,32 @@ export abstract class Client {
   }
 
   notify(noticeable: Noticeable, name: string | Partial<Notice> = undefined, extra: Partial<Notice> = undefined): boolean {
-    let preConditionError: Error = null
     const notice = this.makeNotice(noticeable, name, extra)
-    if (!notice) {
-      this.logger.debug('failed to build error report')
-      preConditionError = new Error('failed to build error report')
-    }
 
-    if (!preConditionError && this.config.reportData === false) {
-      this.logger.debug('skipping error report: honeybadger.js is disabled', notice)
-      preConditionError = new Error('honeybadger.js is disabled')
-    }
-
-    if (!preConditionError && this.__developmentMode()) {
-      this.logger.log('honeybadger.js is in development mode; the following error report will be sent in production.', notice)
-      preConditionError = new Error('honeybadger.js is in development mode')
-    }
-
-    if (!preConditionError && !this.config.apiKey) {
-      this.logger.warn('could not send error report: no API key has been configured', notice)
-      preConditionError = new Error('missing API key')
-    }
     // we need to have the source file data before the beforeNotifyHandlers,
     // in case they modify them
     const sourceCodeData = notice && notice.backtrace ? notice.backtrace.map(trace => shallowClone(trace) as BacktraceFrame) : null
-    const beforeNotifyResult = runBeforeNotifyHandlers(notice, this.__beforeNotifyHandlers)
-    if (!preConditionError && !beforeNotifyResult) {
-      this.logger.debug('skipping error report: beforeNotify handlers returned false', notice)
-      preConditionError = new Error('beforeNotify handlers returned false')
-    }
+    const preConditionsResult = this.__runPreconditions(notice)
```

Comment:
Big fan of moving the preconditions to their own function! 

I think it would be more intuitive if the `beforeNotify` handlers were not run within the precondition function -- conceptually they are pretty different since the preconditions are defined by us and the `beforeNotify` handlers are defined by the user.

Reply:
🤔 Hmm, the way I thought about this is that `beforeNotify` handlers are a subset of the precondition checks.
And we even have beforeNotify handlers that are defined by us (see `browser.ts` and `server.ts` in the js package).

Reply:
Reasonable either way!